Add support for boxes and templates

shigma opened the pull request:
#### Support built-in characters and character encoding in strings.
```mathematica
  "\[alpha]"
(* ^^^^^^^^ constant.character.built-in.wolfram *)

  "\:123456"
(* ^^^^^^ constant.character.encoding.wolfram *)
```
---
#### Not treat `\` as escape anymore, but special characters like `\n` will be recognized as escapes as usual.
```mathematica
  "This is\n\a string. (* not a comment *)"
(*^ punctualation.defination.string.begin *)
(*        ^^ constant.character.escape *)
(*          ^^ string.quoted *)
```
---
#### Support [string templates](https://reference.wolfram.com/language/ref/StringTemplate.html).
```mathematica
  StringTemplate["Value `a`: <* Range[#n] *>."][<|"a" -> 1234, "n" -> 3|>]
(*                      ^^^ variable.parameter *)
(*                           ^^ keyword.operator.template-expression *)
```
---
#### Support [string representation of boxes](https://reference.wolfram.com/language/tutorial/StringRepresentationOfBoxes.html).
```mathematica
  "box 1: \!\(x\^2\); box 2: \(y\^3\) "
(*        ^ keyword.operator.string-box *)
(*             ^^ keyword.operator.x-scriptBox *)
(*                           ^^^^^^^^ string.quoted *)

  \( \)
(*^^ punctuation.section.box.begin.wolfram *)
(*^^^^^ meta.box.wolfram *)
(*   ^^ punctuation.section.box.end.wolfram *)
```

shigma committed:
add support for boxes and templates

shigma committed:
add some adjustments

shigma committed:
built-in and encoding characters

batracos commented on WolframLanguage.sublime-syntax:
```diff
@@ -105,14 +104,38 @@ contexts:
           scope: constant.character.escape.wolfram
-        - match: (?<=[\s\"\`])\`\w*\`
-          scope: constant.other.placeholder.wolfram
+        - match: \\:\w{4}
```
This only works for hexadecimal numbers (`[0-9A-Fa-f]`) and there is also the `\.nn` syntax for two digits codes. 
```suggestion
        - match: \\\.[0-9A-Fa-f]{2}|\\:[0-9A-Fa-f]{4}
```

batracos commented on WolframLanguage.sublime-syntax:
```diff
@@ -105,14 +104,38 @@ contexts:
+            - include: boxed
+
+        # escape characters
+        - match: \\([nrtb]|\r?\n)
```
I don't understand this: why `\n` and `\r\n` need an extra case? Also, the second part of the regex is never going to match.

If you want to match all the "escapable" characters the group should be `\\[nrtbf\\\"]`, which includes `\f` (new page—I had no idea about this), `\\` and `\"`.

shigma replied:
[This documentation](https://reference.wolfram.com/language/ref/String.html) declares an escaping behavior  before new lines.

batracos replied:
Oh I see your point now! I missed the escaping `\`.
As this behaviour is true in general; I suggest you to add it in the `expressions` and the `comment-block` scopes as well.

shigma replied:
> I suggest you to add it in the `expressions` and the `comment-block` scopes as well.

Sorry, but I don't get your point. Does adding a escaping `\` rule make any difference?

batracos replied:
Forget about the first bit. I just misread.
The second part refers to the fact that the newline escaping is working also in comments and normal code, therefore we should add the `\\\r?\n` pattern in more scopes. Try pasting this in a notebook.
```
{1, 2, \
3}

(*{1, 2, \
3}*)

"{1, 2, \
3}"
```

shigma replied:
In that case, how about a prototype of an empty match `\\\r?\n`?

batracos replied:
Empty match?

http://geek-and-poke.com/geekandpoke/2014/2/5/babylon

shigma replied:
> Empty match?

Well, actually I mean a match without any `scope`.

batracos commented on WolframLanguage.sublime-syntax:
```diff
@@ -105,14 +104,38 @@ contexts:
         - match: \"
```
I would move this line at the and to avoid popping when an escaped `"` occurs.

shigma replied:
If escape characters include `"`, there is no need to change here.

batracos replied:
There are still some issues though. Compare
```
"\!\(\*FractionBox[1, 2]\)"
"\!\(\*FractionBox[\"1\", 2]\)"
```

shigma replied:
Sounds a difficult problem for me. I haven't worked out a solution.

Perhaps `with_prototype` will work but every time I used this key my sublime crashed.

batracos replied:
It's very corner case. I suggest to make an issue for tracking purpose and not lose too much time now.

batracos commented on WolframLanguage.sublime-syntax:
```diff
@@ -330,3 +352,26 @@ contexts:
+      scope: keyword.operator.FormBox.wolfram
+    - match: \\\*
+      scope: keyword.operator.box-constructor.wolfram
+      push: expressions
```
Why do you need to push the `expressions` scope when it is already included?

batracos commented on WolframLanguage.sublime-syntax:
```diff
@@ -105,14 +104,38 @@ contexts:
+          scope: keyword.operator.string-box.wolfram
+          push:
+            # workaround plain scope in string meta scope
+            - meta_scope: string.unquoted.yaml meta.string-box.wolfram
```
Why the YAML scope name here?

shigma replied:
Just to work around a plain scope in string meta scope, otherwise the "expression" inside the context will be colored yellow.

I'll try to remove the meta scope `string.quoted.wolfram` here to prevent the behavior.

Any better practice?

batracos replied:
I honestly have no idea.

shigma replied:
In that case, I have to remove the meta scope instead.

batracos commented on WolframLanguage.sublime-syntax:
```diff
@@ -105,14 +104,38 @@ contexts:
+            - meta_scope: string.unquoted.yaml meta.template-expression.wolfram
+            - match: \*>
+              scope: keyword.operator.template-expression.wolfram
+              pop: true
```
You also need to match `(?=\")` to close unbalanced scopes.

batracos commented on syntax_test_wolfram_language.wl:
```diff
@@ -6,129 +6,129 @@
+  11.11`
+(*^^^^^^ constant.numeric *)
+  11.11`11.11
 (* ^^^^^^^^^^^ constant.numeric *)
```
Now there's an extra `^` here.

batracos commented on syntax_test_wolfram_language.wl:
```diff
@@ -148,125 +148,135 @@
-(*  ^ variable.parameter*)
-(*      ^^^^^^^^^^^^^^ variable.function*)
-(*                        ^^ keyword.operator*)
+(* ^ entity.name.function *)
```
Wrong alignment here and in the four below.

batracos commented on syntax_test_wolfram_language.wl:
```diff
@@ -148,125 +148,135 @@
-(*                       ^ string.quoted *)
-(*                                    ^ constant.character.escape *)
+  "This is\n\a string. (* not a comment *)"
+(*^ punctualation.defination.string.begin *)
```
Typo

```suggestion
(*^ punctuation.definition.string.begin *)
```

batracos committed:
optimize encoding pattern

Co-Authored-By: Shigma <33423008+Shigma@users.noreply.github.com>

shigma committed:
update string syntax

batracos committed:
fix a typo

Co-Authored-By: Shigma <33423008+Shigma@users.noreply.github.com>

batracos commented on syntax_test_wolfram_language.wl:
```diff
@@ -185,88 +184,100 @@
-(*^^^ meta.parts.wolfram punctuation.section.parts.begin.wolfram *)
-(*   ^ meta.parts.wolfram *)
-(*    ^^ meta.parts.wolfram punctuation.section.parts.end.wolfram *)
+  [[ ]]
```
The space was there on purpose.

batracos commented on syntax_test_wolfram_language.wl:
```diff
@@ -281,6 +292,8 @@ Block[
+ *)
 
-(* ::s:: *)
+(* ::Input:: *)
```
Can you move this to the `COMMENTS` section?

batracos commented on syntax_test_wolfram_language.wl:
```diff
@@ -148,34 +148,33 @@
-(*      ^^^^^^^^^^^^^^ variable.function*)
-(*                        ^^ keyword.operator*)
+(*^ entity.name.function *)
+(* ^ variable.parameter *)
```
Only the first test was wrong. Now the parameter is not matching (same in the ones below).

batracos replied:
`variable.parameter` is not aligned with `x_`.
Same issue at line 173 and 177

shigma committed:
add some adjustments

chere005 commented:
I think the examples look good and haven't immediately found any issues. @Shigma I'll leave this on in my editor for a day or so and this should be an easy approval

chere005 commented:
Thanks for the efforts @Shigma and @batracos, these updates have all looked nice, sorry I haven't been too responsive

chere005 commented:
At first glance it looked like the escape \ fix was correct, but now I'm not so sure:

<img width="782" alt="screen shot 2018-11-08 at 1 17 59 pm" src="https://user-images.githubusercontent.com/28164815/48221914-ec8b9d00-e358-11e8-8262-76a13b7fd6e1.png">

chere005 commented:
This breaks strings like ```"FOO`BAR`BAZ`"```, which is a big regression that must be fixed before this is merged

Shigma commented:
> This breaks strings like `` "FOO`BAR`BAZ`" `` where this represents a context, which is a big regression that must be fixed before this is merged

Emmm, it does. How do you plan to solve it?

Shigma commented:
> This breaks strings like `` "FOO`BAR`BAZ`" `` where this represents a context, which is a big regression that must be fixed before this is merged

Emmm, it does. How do you plan to solve it?

Shigma commented:
> I'm not sure I see a difference.

What do you mean by "difference"?

Shigma commented:
> At first glance it looked like the escape \ fix was correct, but now I'm not so sure:
> 
> <img alt="screen shot 2018-11-08 at 1 17 59 pm" width="782" src="https://user-images.githubusercontent.com/28164815/48221914-ec8b9d00-e358-11e8-8262-76a13b7fd6e1.png">

It looks good to me.

shigma committed:
add newline prototype

chere005 commented:
> > I'm not sure I see a difference.
> 
> What do you mean by "difference"?

When I pasted these code snippets and applied your changes, the highlighting didn't change at all.

Shigma commented:
> > > I'm not sure I see a difference.
> > 
> > 
> > What do you mean by "difference"?
> 
> When I pasted these code snippets and applied your changes, the highlighting didn't change at all.

It's wield. Syntax rules before this PR should not have been able to handle these syntaxes. Can you tell me which commit?

Shigma commented:
> I'm not sure, this is a much bigger issue than not supporting StringTemplate. I discussed this with @batracos at one point back in May actually and we couldn't find a good solution.. The closest thing I came up with was something along these lines for L140:
> 
> ```
> (?<=[^a-zA-Z0-9$]|[\"\`])\`\w*\`|\`\w*\`(?=[^a-zA-Z0-9$"])
> ```
> It seems to work but @batracos didn't like it. @Shigma and @batracos can we improve this, use it, or come up with something better? If not, we should hold off on StringExpression backtick support

I'm not sure I get your point by ``(?<=[^a-zA-Z0-9$]|[\"\`])`` but the thing is slots in a StringTemplate can be placed just between letters:

```mathematica
StringTemplate["FOO`BAR`BAZ`"][<|"BAR" -> " - "|>]
Out[11]= "FOO - BAZ`"
```
---
The syntax for python didn't work out a solution for string templates, either:
```python
str = 'I just want a {plain} brace'
                   # ^^^^^^^ constant.other.placeholder.python
```

We don't show an attempt to `format` it and it appears as if we do.

---
If you really want to distinguish those behaviour, maybe we can just match the function outside the string, for example, treat strings inside a `Begin` function as "plain" to some extent. I think provide different functions with different inner context is a good solution to all the problems like this, and functions like `Block` has applied this solution before.

I'm interested in it but suggest not to do such things in this PR but open a new one instead because this PR has already does too many things.

Shigma commented:
> Clearly back slash on its own isn't supported from the second output, and when mathematica doesn't complain it's a bug. We should always highlight the character after \ in a string to indicate there is an escape happening, this is crucial to help visually indicate bugs.

Mathematica is not that strict:
```mathematica
In[21]:= "\a" // InputForm
(* Syntax::stresc: Unknown string escape \a. *)
Out[21]//InputForm= "\\a"

In[31]:= "\a"
Out[31]= "\\a"
```

Should out program be that strict? I just don't know.

If yes, maybe we can preserve the current rule and append the following rule:
```yaml
match: \\[\s\S]
scope: invalid.character.escape.wolfram
```

chere005 commented:
> > Clearly back slash on its own isn't supported from the second output, and when mathematica doesn't complain it's a bug. We should always highlight the character after \ in a string to indicate there is an escape happening, this is crucial to help visually indicate bugs.
> 
> Mathematica is not that strict:
> 
> ```mathematica
> In[21]:= "\a" // InputForm
> (* Syntax::stresc: Unknown string escape \a. *)
> Out[21]//InputForm= "\\a"
> 
> In[31]:= "\a"
> Out[31]= "\\a"
> ```
> 
> Should out program be that strict? I just don't know.
> 
> If yes, maybe we can preserve the current rule and append the following rule:
> 
> ```yaml
> match: \\[\s\S]
> scope: invalid.character.escape.wolfram
> ```

I think you missed my point. Mathematica IS that strict. If the message isn't appearing, it is a bug that it isn't appearing. That syntax is never actually valid, and we must indicate that `\` followed by any char means it's an attempted escape. I'm fine, and even encourage giving these a different different colors (valid vs invalid escapes) but `\` on its own should never be unformatted

chere005 commented:
> > I'm not sure, this is a much bigger issue than not supporting StringTemplate. I discussed this with @batracos at one point back in May actually and we couldn't find a good solution.. The closest thing I came up with was something along these lines for L140:
> > ```
> > (?<=[^a-zA-Z0-9$]|[\"\`])\`\w*\`|\`\w*\`(?=[^a-zA-Z0-9$"])
> > ```
> > It seems to work but @batracos didn't like it. @Shigma and @batracos can we improve this, use it, or come up with something better? If not, we should hold off on StringExpression backtick support
> 
> I'm not sure I get your point by `` (?<=[^a-zA-Z0-9$]|[\"\`]) `` but the thing is slots in a StringTemplate can be placed just between letters:
> 
> ```mathematica
> In[11]:= StringTemplate["FOO`BAR`BAZ`"][<|"BAR" -> "-"|>]
> Out[11]= "FOO-BAZ`"
> ```
> 
> The syntax for python didn't work out a solution for string templates, either:
> 
> ```python
> str = 'I just want a {plain} brace'
>                    # ^^^^^^^ constant.other.placeholder.python
> ```
> 
> We don't show an attempt to `format` it and it appears as if we do.
> 
> If you really want to distinguish those behaviour, maybe we can just match the function outside the string, for example, treat strings inside a `Begin` function as "plain" to some extent. I think provide different functions with different inner context is a good solution to all the problems like this, and functions like `Block` has applied this solution before.
> 
> I'm interested in it but suggest not to do such things in this PR but open a new one instead because this PR has already does too many things.

Contexts show up in many other strings. We **absolutely cannot break this** for StringTemplated strings. All strings with backticks are valid containing backticks. Only in the case for StringTemplate are the backticks special. What this means is assume and default to the backtick being a normal backtick in a string. I was suggesting a hacky solution so that sometimes StringTemplate would be right but that contexts would always be right, but I'm less convinced by this. I 100% think we should not support StringTemplate backtick syntax, please remove it from this PR unless there is a much, much better solution.

shigma committed:
remove template recognize and optimize escaping behaviour

1. Now string templates will only be colored in assignment of usages and inside StringTemplate and TemplateApply.
2. Support 3-octal encoding.
3. Better detection for invalid strings.

shigma commented:
> Contexts show up in many other strings. We **absolutely cannot break this** for StringTemplated strings. All strings can contain normal backticks. Only in the case for StringTemplate are the backticks special. What this means is assume and default to the backtick being a normal backtick in a string, unless we can properly distinguish. I was suggesting a hacky solution so that sometimes StringTemplate would be right but that contexts would always be right, but I'm less convinced by this. I 100% think we should not support StringTemplate backtick syntax, please remove it from this PR unless there is a much, much better solution.

I don't entirely agree with you but given that the colorization for string template is a break change I decide to remove all the related rules from general string recognition but preserve them for some functions (specifically StringTemplate and TemplateApply) and assignment for usages (which will automatically use string template when displayed in a [message](https://reference.wolfram.com/language/ref/Message.html)).

Is this OK?

shigma committed:
tiny fix

shigma commented:
> I think you missed my point. Mathematica IS that strict. If the message isn't appearing, it is a bug that it isn't appearing. That syntax is never actually valid, and we must indicate that `\` followed by any char means it's an attempted escape. I'm fine, and even encourage giving these a different different colors (valid vs invalid escapes) but `\` on its own should never be unformatted

I see. Hope the following rules are satisfactory:
https://github.com/ViktorQvarfordt/Sublime-WolframLanguage/blob/4fa3e82dd4f677e88056e9f50f4093d74a00785c/WolframLanguage.sublime-syntax#L137-L161

I found only `[a-zA-Z]` except `[nrtbf]` can result in a error and ``[()"!^%&+_*@`/\\]`` after a back-slant will be recognized as a special character. Under other circumstances, any character after a back-slant will be recognized as usual:
```mathematica
In[69]:= "\8" // Characters
Out[69]= {"\\", "8"}
```

chere005 commented:
> > Contexts show up in many other strings. We **absolutely cannot break this** for StringTemplated strings. All strings can contain normal backticks. Only in the case for StringTemplate are the backticks special. What this means is assume and default to the backtick being a normal backtick in a string, unless we can properly distinguish. I was suggesting a hacky solution so that sometimes StringTemplate would be right but that contexts would always be right, but I'm less convinced by this. I 100% think we should not support StringTemplate backtick syntax, please remove it from this PR unless there is a much, much better solution.
> 
> I don't entirely agree with you but given that the colorization for string template is a break change I decide to remove all the related rules from general string recognition but preserve them for some functions (specifically StringTemplate and TemplateApply) and assignment for usages (which will automatically use string template when displayed in a [message](https://reference.wolfram.com/language/ref/Message.html)).
> 
> Is this OK?

This is a great idea, I'll test it soon.

chere005 commented:
> > Contexts show up in many other strings. We **absolutely cannot break this** for StringTemplated strings. All strings can contain normal backticks. Only in the case for StringTemplate are the backticks special. What this means is assume and default to the backtick being a normal backtick in a string, unless we can properly distinguish. I was suggesting a hacky solution so that sometimes StringTemplate would be right but that contexts would always be right, but I'm less convinced by this. I 100% think we should not support StringTemplate backtick syntax, please remove it from this PR unless there is a much, much better solution.
> 
> I don't entirely agree with you but given that the colorization for string template is a break change I decide to remove all the related rules from general string recognition but preserve them for some functions (specifically StringTemplate and TemplateApply) and assignment for usages (which will automatically use string template when displayed in a [message](https://reference.wolfram.com/language/ref/Message.html)).
> 
> Is this OK?

I want to bring the idea of if there are spaces around the backticks back on the table. While it won't fix every other StringTemplate, it would fix most of the ones that I use. There would just be some instances (when the backticks where next to other characters but it was actually a StringTemplate and not inside the functions we whitelisted.. so very rare). I also suggest we add Success and Failure to the list of functions which commonly use StringTemplate and always turn on the syntax highlighting for this.

chere005 commented:
> > I think you missed my point. Mathematica IS that strict. If the message isn't appearing, it is a bug that it isn't appearing. That syntax is never actually valid, and we must indicate that `\` followed by any char means it's an attempted escape. I'm fine, and even encourage giving these a different different colors (valid vs invalid escapes) but `\` on its own should never be unformatted
> 
> I see. Hope the following rules are satisfactory:
> 
> [Sublime-WolframLanguage/WolframLanguage.sublime-syntax](https://github.com/ViktorQvarfordt/Sublime-WolframLanguage/blob/4fa3e82dd4f677e88056e9f50f4093d74a00785c/WolframLanguage.sublime-syntax#L137-L161)
> 
> Lines 137 to 161 in [4fa3e82](/ViktorQvarfordt/Sublime-WolframLanguage/commit/4fa3e82dd4f677e88056e9f50f4093d74a00785c)
>      # escape characters 
>      - match: \\[-"nrtbf()!^%&+_*@`/\\] 
>        scope: constant.character.escape.wolfram 
>      - match: |- 
>          (?x)( 
>            \\[0-7]{3}| 
>            \\\.[0-9A-Fa-f]{2}| 
>            \\:[0-9A-Fa-f]{4} 
>          ) 
>        scope: constant.character.encoding.wolfram 
>      - match: \\\[({{named_characters}})\] 
>        scope: constant.character.built-in.wolfram 
>   
>      # invalid characters 
>      - match: |- 
>          (?x)( 
>            \\[0-7]{1,2}(?=[^0-7])| 
>            \\\.[0-9A-Fa-f]?(?=[^0-9A-Fa-f])| 
>            \\:[0-9A-Fa-f]{0,3}(?=[^0-9A-Fa-f]) 
>          ) 
>        scope: invalid.character.encoding.wolfram 
>      - match: \\\[\w+\] 
>        scope: invalid.character.built-in.wolfram 
>      - match: \\[a-zA-Z] 
>        scope: invalid.character.escape.wolfram 
> 
> I found only `[a-zA-Z]` except `[nrtbf]` can result in a error and `` [()"!^%&+_*@`/\\] `` after a back-slant will be recognized as a special character. Under other circumstances, any character after a back-slant will be recognized as usual:
> 
> ```mathematica
> In[69]:= "\8" // Characters
> Out[69]= {"\\", "8"}
> ```

At first glance this seems to work, but it does look like it can be cleaned up. Perhaps another thing to put an issue for, but I'll take another look next week

chere005 commented:
To be honest I never work with boxes so I'm just not sure if this is expected:

Current Release:
<img width="481" alt="currentrelease" src="https://user-images.githubusercontent.com/28164815/48298958-8d7a7500-e48b-11e8-90d3-988854784c03.png">


This PR's latest commit:
<img width="482" alt="withchanges" src="https://user-images.githubusercontent.com/28164815/48298962-98cda080-e48b-11e8-8d64-24d60a05a755.png">

chere005 commented:
It looks a bit to me like some of the escape character rules are leaking here. I wasn't really sure why this was the list of escape characters, in particular ()%+_@/ could you give some insight on everything past f in the list perhaps?


```[nrtbf()"!^%&+_*@`/\\]```


This is the best resource I've found: https://reference.wolfram.com/language/tutorial/InputSyntax.html apparently \000 is a thing too..

chere005 commented:
I tried to respond to most questions tonight, I won't be back on over the weekend. I plan to leave your most recent commits in my local sublime for a few days at work next week before approving so I have time to see if I notice anything with my normal workflows.

shigma commented:
> To be honest I never work with boxes so I'm just not sure if this is expected:
> 
> Current Release:
> <img alt="currentrelease" width="481" src="https://user-images.githubusercontent.com/28164815/48298958-8d7a7500-e48b-11e8-90d3-988854784c03.png">
> 
> This PR's latest commit:
> <img alt="withchanges" width="482" src="https://user-images.githubusercontent.com/28164815/48298962-98cda080-e48b-11e8-8d64-24d60a05a755.png">

Maybe I worked with boxes before and just forgot all about it later on ...

shigma commented:
> It looks a bit to me like some of the escape character rules are leaking here. I wasn't really sure why this was the list of escape characters, in particular ()%+_@/ could you give some insight on everything past f in the list perhaps?
> 
> `` [nrtbf()"!^%&+_*@`/\\] ``
> 
> This is the best resource I've found: https://reference.wolfram.com/language/tutorial/InputSyntax.html apparently \000 is a thing too..

See this:
```mathematica
In[153]:= Characters /@ {"\+", "\-", "\>"}
Out[153]= {{"\+"}, {"\\", "-"}, {}}
```

shigma commented:
@chere005 The following code may better depict the escaping behavior:
```mathematica
Quiet @ Last @ Reap[
    Scan[
        Sow[#, Check[Length @ Characters @ ToExpression["\"\\" <> # <> "\""], -1]] &,
        CharacterRange[33, 126]
    ],
    _,
    #1 -> StringJoin[#2] &
]
```
With the following result:
- **errored:** `.01234567:ABCDEFGHIJKLMNOPQRSTUVWXYZ[acdeghijklmopqsuvwxyz`
- **escaped:** ``!"%&()*+/@\^_`bfnrt``
- **non-escaped:** `#$',-89;=?]{|}~`
- **disappeared:** `<>`

shigma committed:
optimize escape pattern

shigma committed:
support for regular expression

shigma commented:
Any questions @batracos ?

batracos commented:
Sorry for the long time off the project. I simply did not have the bandwidth to spare.
Except for those three tests not passing this seems good to go.

chere005 commented:
Hold off on this until I can take another look..

Sent from my iPhone

> On Mar 23, 2019, at 12:04 PM, batracos <notifications@github.com> wrote:
> 
> Sorry for the long time off the project. I simply did not have the bandwidth to spare.
> Except for those three tests not passing this seems good to go.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

chere005 commented:
> 
> 
> @chere005 The following code may better depict the escaping behavior:
> 
> ```mathematica
> Quiet @ Last @ Reap[
>     Scan[
>         Sow[#, Check[Length @ Characters @ ToExpression["\"\\" <> # <> "\""], -1]] &,
>         CharacterRange[33, 126]
>     ],
>     _,
>     #1 -> StringJoin[#2] &
> ]
> ```
> 
> With the following result:
> 
>     * **errored:** `.01234567:ABCDEFGHIJKLMNOPQRSTUVWXYZ[acdeghijklmopqsuvwxyz`
> 
>     * **escaped:** `` !"%&()*+/@\^_`bfnrt ``
> 
>     * **non-escaped:** `#$',-89;=?]{|}~`
> 
>     * **disappeared:** `<>`

@batracos and I discussed this, in a string the only characters that should show up as valid escape characters is `nrtbf`  Things like `"\*"` which don't complain end up with a string `"\\*"` which means the `*` was not actually escaped. Please make this change and fix the tests that @batracos mentioned are failing before I make another pass at reviewing this.

chere005 commented:
@Shigma any plans to update this with the changes we mentioned?
